src/runtime/set.cpp

```diff
@@ -100,21 +109,39 @@ BoxedSet* makeNewSet(BoxedClass* cls, Box* container) {
     BoxedSet* rtn = new (cls) BoxedSet();
 
     if (container) {
-        for (Box* e : container->pyElements()) {
-            rtn->s.insert(e);
+        if (PyAnySet_Check(container)) {
+            for (auto&& elt : ((BoxedSet*)container)->s) {
+                rtn->s.insert(elt);
+            }
+        } else if (PyDict_CheckExact(container)) {
+            for (auto&& elt : ((BoxedDict*)container)->d) {
+                rtn->s.insert(elt.first);
+            }
+        } else {
+            for (auto elt : container->pyElements()) {
+                rtn->s.insert(elt);
+            }
         }
     }
     return rtn;
 }
 
-Box* frozensetNew(Box* _cls, Box* container) {
+Box* frozensetNew(Box* _cls, Box* container, BoxedDict* kwargs) {
     RELEASE_ASSERT(_cls->cls == type_cls, "");
     BoxedClass* cls = static_cast<BoxedClass*>(_cls);
     RELEASE_ASSERT(isSubclass(cls, frozenset_cls), "");
 
     // Some optimizations from CPython: frozensets can be shared:
     if (_cls == frozenset_cls) {
-        if (!container) {
+        if (!_PyArg_NoKeywords("frozenset()", kwargs)) {
+            throwCAPIException();
+        }
+        bool isEmpty = false;
+        static BoxedString* len_str = internStringImmortal("__len__");
+        if (container && PyObject_HasAttr(container, len_str)) {
+            isEmpty = PyObject_Size(container) > 0 ? false : true;
+        }
+        if (!container || isEmpty) {
```

Comment:
It looks like cpython's logic (from `frozenset_new`) is to create a frozenset object from the container, but if the resulting frozenset is empty, to discard it and return the emptyfrozenset singleton instead.